### DISCUSSION TITLE
zwave-js-ui, nginx, HA websocket issue

### DISCUSSION THREAD

**Discussion by kantlivelong** [zwave-js/zwave-js-ui#2784] [Q&A]
This is a bit lengthy but here it goes...

I've got a seemingly weird issue with zwave-js-ui and nginx. For whatever reason the websocket connection only works if I CTRL+F5(no-cache reload). If I do a simple F5 refresh afterwards the websocket connection fails.

Digging into this further I've found that on a regular page load/refresh zwave-js-ui or socket.io is returning what seems to be bogus sid and it's always the same sid. If I no-cache refresh a valid sid is returned and the websocket session connects without issue.

I've checked the issue trackers for zwave-js-ui and socket.io as well as done some general web searching to see what I can find. From what I can tell my nginx config should be fine as it does work via CLI via `wscat` and does work on a no-cache reload.

My setup is pretty simple but is as follows
- zwave-js-ui is running inside a container on a raspberry pi with the dongle. Sitting along side it is another container running zigbee2mqtt with it's respective dongle.
- nginx is on the same subnet but running in another machine(VM)
- HA is also running in this environment but is on it's own VM. 

Config: https://pastebin.com/YkytXi15

Checking the Firefox dev console on a regular load/reload I see the first request to socket.io via `https://homeassistant.FQDN:8124/_/zwavejs2mqtt/socket.io/?EIO=4&transport=polling&t=OGVHQXK` and it receives the response `0{"sid":"QHEjB7LIw5v_fO0rAAEl","upgrades":["websocket"],"pingInterval":25000,"pingTimeout":20000,"maxPayload":1000000}`. The sid in this response is _**always**_ the same.

When I no-cache reload I see the same initial connection but with a valid sid. I can verify the same behavior with `curl`

### Mid ticket update:

So I've found that zwave-js-ui will work fine _**until**_ I sign into HA. Once I sign into HA the problematic behavior occurs.

HA URL: https://homeassistant.FQDN:8124/
zwave-js-ui URL: https://homeassistant.FQDN:8124/_/zwavejs2mqtt/

I found that simply adding an extra port (8125) to the nginx config then connecting to zwave-js-ui works perfectly at the URL above _**until**_  I sign into HA at the URL above. 

Is there some sort of cookie or session storage collision going on here?

**Comment by robertsLando** [comment]
Never seen such kind of problems, just to be clear, are you speaking about the zwave-js UI websocket or the home hassistant websocket? Is your problem on the z-ui UI side or home assistant side?

**Comment by kantlivelong** [comment]
The UI Websocket. I have the HA websocket disabled.

**Comment by robertsLando** [comment]
At first sight it seems to me a cache problem in nginx configuration, I'm not an nginx expert but I know hass addon works without any problem (and it's using nginx) and I also have a [working example](https://github.com/zwave-js/zwave-js-ui/tree/master/docker/nginx) using nginx

**Comment by AaronSeibert** [comment]
Your location / matches all requests, and you have websocket headers in there. If you don't need the websocket on / and only on the zigbee/hass/zwavejs parts can probably take it out of there.

FWIW, it may be easier to use a separate host header for each. Would make the configs easier to troubleshoot vs having child paths.

**Reply by kantlivelong** [reply]
So I've finally had some time to look into this a bit more. What I'm seeing is that the websocket connection is being routed to the appropriate backend(zwavejsui). I confirmed this by opening the UI and follow nginx logs. During the storm of HTTP 400 `Session ID unknown` errors I stopped zwavejsui and starting receiving HTTP 502s. 

This brings me right back to the initial ticket where zwavejsui for whatever reason is providing an invalid sid.

**Comment by kantlivelong** [comment]
Okay wow this is REALLY weird....

I ran a packet capture on the host running zwavejsui and on the client desktop machine.

capture from host running zwavejsui:
![image](https://user-images.githubusercontent.com/1585573/208581805-d4bbe827-4a0c-40e8-9f84-18faebae92d2.png)
zwavejsui returns sid `-aj3FwnPgnv_OOZHAAj7`

capture from desktop running firefox:
![image](https://user-images.githubusercontent.com/1585573/208582328-37b5465b-d9ce-4412-a214-5cabebc2067e.png)
Also seeing sid `-aj3FwnPgnv_OOZHAAj7`

Now here's where it all falls on it's face.
![image](https://user-images.githubusercontent.com/1585573/208583315-cb3f1632-6cc8-41df-873d-c9c1f33ee5ea.png)

Somehow the browser is reporting a different response!? What the...

**Comment by kantlivelong** [comment]
I've pinpointed the issue and opened a ticket with Home Assistant - Frontend. In short it's service-worker is caching any http responses with a status of 200. Since socket.io provides the session id in a response with status 200 the data is cached and the correct sid is never recieved by zwavejsui unless the browser is refreshed with the cache being cleared (Shift+F5)

https://github.com/home-assistant/frontend/issues/14836

**Reply by robertsLando** [reply]
@kantlivelong Great finding! Let me know if there could be something to do on my side, keep me updated on this
